### ISSUE TITLE
Connection refused in init scripts

### ISSUE THREAD

**Issue by yarons** [bitnami/bitnami-docker-keycloak#41]
**Description**

Init scripts are loaded before the product so certain changes cannot be done with .sh scripts, only using JSON.

**Steps to reproduce the issue:**

1. Add sh script with kcadm.sh
2. Start docker-compose
3. Observe the "Connection refused" messages in docker-logs.

**Describe the results you received:**

The connection to the service was refused.

**Describe the results you expected:**

The changes should have been made.

**Additional information you deem important (e.g. issue happens only occasionally):**

Adding a pre scripts and post script would probably solve that.

**Version**

- Output of `docker version`:

```
20.10.11
```

- Output of `docker info`:

```
irrelevant.
```

- Output of `docker-compose version` (if applicable):

```
2.2.2
```

**Additional environment details (AWS, VirtualBox, Docker for MAC, physical, etc.):**
irrelevant.

**Comment by javsalgar** [comment]
Hi,

Could it be that the service is stopped? Could you try starting it and then performing the operations?

**Comment by yarons** [comment]
Nope, the service is not up yet, kcadm.sh cannot work without the service responding.

This is the script I'm currently trying:

```bash
#!/bin/bash
(printf "%s" "waiting for ServerXY ..."
while ! timeout 2 echo > /dev/tcp/localhost/8080 &> /dev/null
do
    printf "%c" "."
done
cd /opt/bitnami/keycloak/bin
kcadm.sh config credentials --user admin --password admin --server http://localhost:8080/auth --realm master
kcadm.sh create clients -s clientId=invitease -s 'redirectUris=["http://invitease.localhost/*"]' -s clientAuthenticatorType=client-secret -s secret=d0b8122f-8dfb-46b7-b68a-f5cc4e25d000
kcadm.sh update realms/master -s attributes.frontendUrl=http://keycloak.localhost/auth) &
```

And I'm seeing the following error:

```
invitease-keycloak-1  | Failed to create config file: /opt/bitnami/keycloak/bin/?/.keycloak/kcadm.config
invitease-keycloak-1  | No server specified. Use --server, or 'kcadm.sh config credentials or connection'.
invitease-keycloak-1  | No server specified. Use --server, or 'kcadm.sh config credentials or connection'.
```

I chmoded the script with a+x so it's not the problem.
Some further explanation:
Assuming that the server is not already running I need to wait for it (querying the port until it becomes available).
So I figured that I can simply create a loop inside a background process (using & after the parenthesis) and when the port will become available the kcadm.sh will do its magic, the commands are running but it seems that there are no permissions, any thoughts?

**Comment by javsalgar** [comment]
I believe that the issue might be related to the keycloak server not being up at that moment. You may need to start it before doing that wait. Could you try that?

**Comment by yarons** [comment]
I can't, this is how the docker works, the init scripts are loaded before Keycloak is even running, surprisingly enough this is also the situation with the official JBoss docker.

I have a working solution over here:
https://github.com/Maagan-Michael/invitease/blob/main/init/keycloak/kc-seed.sh

So what happens is that the init scripts are loaded before this script is even launched so kcadm.sh doesn't work, I can't launch it afterwards because this is not how the init directory works.
What I had to do was to wait for port 9990 to become available and then launch my commands on port 8080, waiting for 8080 has no effect because the reverse proxy is available once the docker is up so this condition will always be true.

Since there are several listening ports I couldn't really understand how this whole mecahnism works, I just know that the main process is the standalone.sh script which launches Keycloak eventually and kcadm.sh only works after port 9990 becomes available.

Another thing that happens due to this mechanism is that the service seems available although it doesn't so if I'm using `depends_on` against this Docker, my app will launch without actually waiting and fail until Keycloak will end loading, we've actually added `restart: unless-stopped` to bypass this limitation.

Thanks and sorry for the long story.

**Comment by miguelaeh** [comment]
Hi @yarons ,

The init scripts are executed after the keycloak initialization, but as you stated, after the initialization keycloak is not actually running, just configured to be started later.
If your script requires keylock to be running, you can start it before the required commands, and stop it at the end. The container logic will take care of starting it again when needed using the `run.sh` script.

> Another thing that happens due to this mechanism is that the service seems available although it doesn't so if I'm using depends_on against this Docker, my app will launch without actually waiting and fail until Keycloak will end loading, we've actually added restart: unless-stopped to bypass this limitation.

This is a different thing. That is how docker-compose works, the `depend_on` will mark the container as running regardless of whether the service the container contains is ready to accept connections or not.

**Comment by yarons** [comment]
Hey @miguelaeh,
Correct me if I'm wrong but although there are certain tasks I would like to preform before the service is running there are things which I would like to change against Keycloak (Add a client, change the theme localization, etc.) which I can't do without it so I was wondering whether it's possible to include a preinit and postinit options, this way I won't have to load the service manually before the container does that for me.

BTW If I'll load the service using the init script, what will happen later? The container scripts will try to load it again afterwards.

Regarding `depends_on`: I know how it works, I'm just saying that if the Docker would load Keycloak as PID 1 the container loading time will match the availability of the service, what happens now is that Keycloak container appears as running but the dependant services are crashing or failing because it doesn't actually respond.

**Comment by miguelaeh** [comment]
Hi @yarons ,
> Correct me if I'm wrong but although there are certain tasks I would like to preform before the service is running there are things which I would like to change against Keycloak (Add a client, change the theme localization, etc.) which I can't do without it so I was wondering whether it's possible to include a preinit and postinit options, this way I won't have to load the service manually before the container does that for me.

Currently, the containers do not support such feature. You can either send a PR supporting the execution of custom scripts before initialization or extend the image with your custom logic.

> BTW If I'll load the service using the init script, what will happen later? The container scripts will try to load it again afterwards.

You would need to start and stop the service, then it won't conflict with the normal scripts, since it will be stoped.

> Regarding depends_on: I know how it works, I'm just saying that if the Docker would load Keycloak as PID 1 the container loading time will match the availability of the service, what happens now is that Keycloak container appears as running but the dependant services are crashing or failing because it doesn't actually respond.

I am not fully sure of that, probably even running keycloak as PID 1 the same could happen. For example, while keycloak is starting it would be detected as running, but actually it would be starting. That solution would decrease the times this happens but won't solve it in all situations. The ideal solution is to wait until keycloak is accepting connections before trying to connect to it.

**Comment by yarons** [comment]
I opened this issue because I think it should support that.
The init script is running before Keycloak so I can preform all kinds of operations such as changing files in the container and modifying some settings but I can't seed Keycloak and this is the most important thing this script should support.

Let's take MySQL and Postgresql for example, both their init script mechanisms are importing the .sql files and running the .sh scripts after the respective database service is loaded, otherwise there will be no seeding of the database.
It makes no sense running the preinit script without beind able to control Keycloak programatically.

Extending the container is a bad practice and it's most deinitely an example for a bad design, why would I need to extend a Docker just to seed it? It should be a builtin function (Like most other services do).

Waiting for the process to finish is exactly what depends on was designed for and if the service appears as running but it's not actually running then what is this Docker good for? Doing something that it was designed to do but doesn't do it?

Returning to our MySQL and Postgresql example, if I'll specify `depends_on` for any of these I can be pretty sure that my designated container that depends on them will load whenever these containers are ready, otherwise this setting is useless, it was designed exactly to avoid such cases, implementing an additional try catch logic for a bad design using an extended Docker is a bad practice covering a bad design.

I understand the technology and I understand all the logic but I've been working with Dockers for several years now for many different use cases and this is probably the first time I'm running into such case, if there's a supporting logic for you container it's probably there for a reason and if everyone else implemented init scripts in a certain way this is probably the way to go, I truly don't understand why I need to patch or extend a fully contained service.

Thanks and sorry if I'm too harsh, I'm really trying to understand why is it so.

**Comment by miguelaeh** [comment]
Hi @yarons ,

> Waiting for the process to finish is exactly what depends on was designed for and if the service appears as running but it's not actually running then what is this Docker good for? Doing something that it was designed to do but doesn't do it?

Please check the docker-compose documentation [here](https://docs.docker.com/compose/compose-file/compose-file-v3/#depends_on). It explicitly says the following:

```
depends_on does not wait for db and redis to be “ready” before starting web - only until they have been started. If you need to wait for a service to be ready, see Controlling startup order for more on this problem and strategies for solving it.
```

It is important to differentiate between when the container has started (which is what depends_on checks) and when the service deployed by the container has started (which in your case will be keycloak).

Coming back to the issue itself, what Mysql And Postgresql do is mostly what I suggested in a comment above:

> If your script requires keycloak to be running, you can start it before executing the required commands, and stop it at the end. The container logic will take care of starting it again when needed using the run.sh script.

In MySQL and PostgreSQL that is what is done in their logics, check [here](https://github.com/bitnami/bitnami-docker-mysql/blob/bd9541fc24c3d31e432b5b9d776f10e5c0833b08/8.0/debian-10/rootfs/opt/bitnami/scripts/mysql/setup.sh#L33). As you can see, after running the init scripts, we stop MySQL, and the `run.sh` will start it again. In the MySQL case, after the initialize function MySQL will be running, probably not the case of Keycloak, which would need to be changed to be started before executing the init scripts.

**Comment by yarons** [comment]
@miguelaeh That sounds more like it, thanks.

**Comment by yarons** [comment]
It's not solved yet but if you're interested the following [example](https://github.com/Maagan-Michael/invitease/blob/main/init/keycloak/kc-seed.sh) should help you implement seeding script successfully.

**Comment by miguelaeh** [comment]
Hi,
the issue was closed due to inactivity. Would you like to create a PR implementing the suggested changes?

**Comment by miguelaeh** [comment]
Hi @yarons ,
Thank you for the collaboration, what needs to be implemented is what is described at the last paragraph of this comment https://github.com/bitnami/bitnami-docker-keycloak/issues/41#issuecomment-1008671247

**Comment by yarons** [comment]
@miguelaeh I'll sure try that but please notice that the Keycloak's loading time inside the docker is very long, on my computer it's over 2 minutes, so if we'll launch Keycloak, preform the changes, shut down Keycloak and continue with loading it's going to be at least 4 minutes, are you sure that's desirable?

**Comment by miguelaeh** [comment]
Hi @yarons ,
That would be the price to pay for running custom scripts. I can't see right now any other way of doing this without starting Keycloack twice

**Comment by yarons** [comment]
@miguelaeh Hmmmmm there's my way :)
Pre-init and post-init.
Is there a way to load Keycloak and use kcadm.sh without fully loading it? Maybe some minimal mode or maintenance mode?

The only thing I don't understand is what is the current init situation is good for?
Let's presume I want to add a realm to Keycloak using docker-compose and using JSON is an overkill because I need to specify many parameters that doesn't really matter to me, how can I achieve that?
What I'm trying to say is that I can't find a scenario where the current init method is even usable, can you?

**Comment by miguelaeh** [comment]
Hi @yarons ,
But running an init script is not always necessary, right? I mean, the current situation runs keycloak properly, but the case is you want to run an `.sh` script to make some modifications. That will not be always the case.

>Is there a way to load Keycloak and use kcadm.sh without fully loading it? Maybe some minimal mode or maintenance mode?

About this, honestly, I don't know it

**Comment by yarons** [comment]
@miguelaeh  I will try looking at the official docker for answers, thanks :)

**Comment by zhongweili2010** [comment]
I also ran into the same issue trying to seed Keycloak from script. But I believe the issue is simply due to the path to config file is messed up.
There is always a "?" in the path when failed to create config file. Bitnami changed file path from /opt/jboss/keycloak to/opt/bitnami/keycloak, so I suspect a typo somewhere. Could someone with familiarity to Keycloak help review the source code of kcadm.sh?

```
[root@c79-20-99 helm-chart]# kubectl exec -it myrelease-keycloak-0 -n ucp -- /opt/bitnami/keycloak/bin/kcadm.sh config truststore --trustpass mypassword /opt/bitnami/keycloak/certs/keycloak.truststore.jks
Defaulted container "keycloak" out of: keycloak, init-certs (init)
Failed to create config file: /?/.keycloak/kcadm.config
```

**Comment by miguelaeh** [comment]
Hi @zhongweili2010 ,
The source code of that file is the following:
```
#!/bin/sh
case "`uname`" in
    CYGWIN*)
        CFILE = `cygpath "$0"`
        RESOLVED_NAME=`readlink -f "$CFILE"`
        ;;
    Darwin*)
        RESOLVED_NAME=`readlink "$0"`
        ;;
    FreeBSD)
        RESOLVED_NAME=`readlink -f "$0"`
        ;;
    Linux)
        RESOLVED_NAME=`readlink -f "$0"`
        ;;
esac

if [ "x$RESOLVED_NAME" = "x" ]; then
    RESOLVED_NAME="$0"
fi

DIRNAME=`dirname "$RESOLVED_NAME"`


java $KC_OPTS -cp $DIRNAME/client/keycloak-admin-cli-16.1.1.jar org.keycloak.client.admin.cli.KcAdmMain "$@"
```

The actual issue does not look to be there, probably inside the tar.gz called there

**Comment by fmulero** [comment]
We are moving to [a new repository](https://github.com/bitnami/containers) in order to be aligned with the approach followed in [bitnami/charts](https://github.com/bitnami/charts). More information [here](https://blog.bitnami.com/2022/07/new-source-of-truth-bitnami-containers.html). If you are still interested on this feature, please open a new issue in [bitnami/containers](https://github.com/bitnami/containers) repository
